reasoner/planner/CostEstimator.java

```diff
@@ -0,0 +1,417 @@
+        }
+
+        private FunctionalIterator<Resolvable<?>> resolvables() {
+            return iterate(logicMgr.compile(conjunction));
+        }
+
+        private Set<Variable> allVariables() {
+            return iterate(conjunction.pattern().variables())
+                    .filter(v -> v.isThing())
+                    .toSet();
+        }
+
+        private void initializeOnce() {
+            // TODO: Move this after the retrievable estimates, so cycle-head estimates are easy.
+            registerTriggeredRules();
+
+            assert unaryCostCover == null;
+            unaryCostCover = new HashMap<>();
+            {
+                // Create a baseline estimate from type information
+                List<LocalEstimate> unaryEstimates = new ArrayList<>();
+                iterate(allVariables()).forEachRemaining(v -> {
+                    unaryEstimates.add(estimateFromTypes(v));
+                });
+
+                unaryEstimates.sort(Comparator.comparing(x -> x.answerEstimate(this, new HashSet<>(x.variables))));
+                for (LocalEstimate unaryEstimate : unaryEstimates) {
+                    if (!unaryCostCover.containsKey(unaryEstimate.variables.get(0))) {
+                        unaryCostCover.put(unaryEstimate.variables.get(0), new CostCover(unaryEstimate.answerEstimate(this, new HashSet<>(unaryEstimate.variables))));
+                    }
+                }
+            }
+
+            assert multivarEstimates == null;
+            multivarEstimates = new HashMap<>();
+            resolvables().forEachRemaining(resolvable -> multivarEstimates.put(resolvable, new ArrayList<>()));
+
+            //TODO: How does traversal handle type Variables?
+
+            resolvables().filter(Resolvable::isRetrievable).map(Resolvable::asRetrievable).forEachRemaining(retrievable -> {
+                Set<Concludable> concludablesInRetrievable = ResolvableConjunction.of(retrievable.pattern()).positiveConcludables();
+                iterate(concludablesInRetrievable).forEachRemaining(concludable -> {
+                    multivarEstimates.get(retrievable).addAll(computeEstimatesFromConcludable(concludable));
+                });
+            });
+
+            iterate(resolvables()).filter(this::isInferrable).map(Resolvable::asConcludable).forEachRemaining(concludable -> {
+                multivarEstimates.get(concludable).addAll(computeEstimatesFromConcludable(concludable));
+            });
+
+            if (this.negatedsCost == -1) {
+                this.negatedsCost = resolvables().filter(Resolvable::isNegated).map(Resolvable::asNegated)
+                        .flatMap(negated -> iterate(negated.disjunction().conjunctions()))
+                        .map(negatedConjunction -> {
+                            costEstimator.registerConjunction(negatedConjunction);
+                            return costEstimator.estimateAllAnswers(negatedConjunction);
+                        }).reduce(0L, Long::sum);
+            }
```

Comment:
Negations can't contribute to answers (though they do rule them out, but that's out of scope)
They do contribute to the cost of evaluating this condition though.